core-parser/src/main/java/org/uniprot/core/parser/fasta/uniparc/UniParcProteomeFastaParser.java

```diff
@@ -0,0 +1,134 @@
+                    hasSourceActive = true;
+                }
+            }
+        }
+        StringBuilder sb = new StringBuilder();
+        if(!uniProtXrefs.isEmpty()){
+            sb.append(getFastaHeader(uniProtXrefs, hasActive, id, proteomeId, false));
+        } else {
+            sb.append(getFastaHeader(sourceXrefs, hasSourceActive, id, proteomeId, true));
+        }
+        sb.append("\n");
+        sb.append(parseSequence(entry.getSequence().getValue()));
+        return sb.toString();
+    }
+
+    private static String getProteomeId(Property property) {
+        String result = null;
+        String[] sourcePropertyValues = property.getValue().split(",");
+        if(sourcePropertyValues.length == 1){
+            String[] propertyValue = sourcePropertyValues[0].split(":");
+            if(propertyValue.length > 1){
+                result = propertyValue[1];
+            }
+        }
+        return result;
+    }
+
+    private static StringBuilder getFastaHeader(List<UniParcCrossReference> xrefs, boolean hasActive, String id, String proteomeId, boolean isSource) {
+        Set<String> proteinName = new LinkedHashSet<>();
+        Set<String> geneNames = new LinkedHashSet<>();
+        Set<String> accessions = new LinkedHashSet<>();
+        Set<String> sourceIds = new LinkedHashSet<>();
+        Set<String> component = new LinkedHashSet<>();
+        Organism organism = null;
+        for(UniParcCrossReference xref: xrefs) {
+            if (xref.isActive() == hasActive) {
+                addOrIgnoreNull(xref.getProteinName(), proteinName);
+                addOrIgnoreNull(xref.getGeneName(), geneNames);
+                organism = xref.getOrganism();
+
+                if (isSource) {
+                    addOrIgnoreNull(xref.getId(), sourceIds);
+                    addOrIgnoreNull(xref.getComponent(), component);
+                } else {
+                    addOrIgnoreNull(xref.getId(), accessions);
+                    if (notNullNotEmpty(xref.getProperties())) {
+                        xref.getProperties().stream()
+                                .filter(p -> UniParcCrossReference.PROPERTY_SOURCES.equals(p.getKey()))
+                                .map(Property::getValue)
+                                .forEach(value -> {
+                                    String[] sources = value.split(",");
+                                    for (String source : sources) {
+                                        String[] ids = source.split(":");
+                                        if (ids.length > 1 && proteomeId.equals(ids[1])) {
+                                            sourceIds.add(ids[0]);
+                                        }
+                                        if (ids.length > 2 && proteomeId.equals(ids[1])) {
+                                            component.add(ids[2]);
+                                        }
+                                    }
+                                });
```

Comment:
Extract the logic for processing xref.getProperties() into a helper method, simplifying the loop body.
We should keep the lambda small. we can extract the `for(source:sources)` logic another method.